trie/zk_trie_node.go

```diff
@@ -0,0 +1,223 @@
+		preImageFlag := int(b[curPos])
+		curPos += 1
+		if preImageFlag != 0 {
+			copy(n.KeyPreimage[:], b[curPos:curPos+32])
+		}
+	case NodeTypeEmpty:
+		break
+	default:
+		return nil, ErrInvalidNodeFound
+	}
+	return &n, nil
+}
+
+// LeafKey computes the key of a leaf node given the hIndex and hValue of the
+// entry of the leaf.
+func LeafKey(k, v *zkt.Hash) (*zkt.Hash, error) {
+	return zkt.HashElems(big.NewInt(1), k.BigInt(), v.BigInt())
+}
+
+// Key computes the key of the node by hashing the content in a specific way
+// for each type of node.  This key is used as the hash of the merklee tree for
+// each node.
+func (n *Node) Key() (*zkt.Hash, error) {
+	if n.key == nil { // Cache the key to avoid repeated hash computations.
+		// NOTE: We are not using the type to calculate the hash!
+		switch n.Type {
+		case NodeTypeMiddle: // H(ChildL || ChildR)
+			var err error
+			n.key, err = zkt.HashElems(n.ChildL.BigInt(), n.ChildR.BigInt())
+			if err != nil {
+				return nil, err
+			}
+		case NodeTypeLeaf:
+			var err error
+			n.valueHash, err = zkt.PreHandlingElems(n.CompressedFlags, n.ValuePreimage)
+			if err != nil {
+				return nil, err
+			}
+
+			n.key, err = LeafKey(n.NodeKey, n.valueHash)
+			if err != nil {
+				return nil, err
+			}
+
+		case NodeTypeEmpty: // Zero
+			n.key = &zkt.HashZero
+		default:
+			n.key = &zkt.HashZero
+		}
+	}
+	return n.key, nil
+}
+
+func (n *Node) ValueKey() (*zkt.Hash, error) {
+	if _, err := n.Key(); err != nil {
+		return nil, err
+	}
+	return n.valueHash, nil
+}
+
+// Data returns the wrapped data inside LeafNode and cast them into bytes
+// for other node type it just return nil
+func (n *Node) Data() []byte {
+	switch n.Type {
+	case NodeTypeLeaf:
+		var data []byte
+		hdata := (*reflect.SliceHeader)(unsafe.Pointer(&data))
+		//TODO: uintptr(reflect.ValueOf(n.ValuePreimage).UnsafePointer()) should be more elegant but only available until go 1.18
+		hdata.Data = uintptr(unsafe.Pointer(&n.ValuePreimage[0]))
+		hdata.Len = 32 * len(n.ValuePreimage)
+		hdata.Cap = hdata.Len
+		return data
+	default:
+		return nil
+	}
+}
+
+// Value returns the value of the node.  This is the content that is stored in
+// the backend database.
+func (n *Node) Value() []byte {
+	switch n.Type {
+	case NodeTypeMiddle: // {Type || ChildL || ChildR}
+		bytes := []byte{byte(n.Type)}
+		bytes = append(bytes, n.ChildL[:]...)
+		bytes = append(bytes, n.ChildR[:]...)
+		return bytes
+	case NodeTypeLeaf: // {Type || Data...}
+		bytes := []byte{byte(n.Type)}
+		bytes = append(bytes, n.NodeKey[:]...)
+		tmp := make([]byte, 4)
+		compressedFlag := (n.CompressedFlags << 8) + uint32(len(n.ValuePreimage))
+		binary.LittleEndian.PutUint32(tmp, compressedFlag)
+		bytes = append(bytes, tmp...)
+		for _, elm := range n.ValuePreimage {
+			bytes = append(bytes, elm[:]...)
+		}
+		if n.KeyPreimage != nil {
+			bytes = append(bytes, 1)
+			bytes = append(bytes, n.KeyPreimage[:]...)
+		} else {
+			bytes = append(bytes, 0)
+		}
```

Comment:
I am not really against it, but I feel like a more commone way is to 
```
		if n.KeyPreimage != nil {
			bytes = append(bytes, n.KeyPreimage[:]...)
		} else {
			// do nothing
		}
```

or 


```
		if n.KeyPreimage != nil {
			bytes = append(bytes, len(n.KeyPreimage[:]))
			bytes = append(bytes, n.KeyPreimage[:]...)
		} else {
			bytes = append(bytes, 0)
		}
```

Reply:
The keyPreimage is supposed to be 32 bytes but the latter way seems better for more flexible. I would try to update on it.